### ISSUE TITLE
Get mirrorlist for a given IP

### ISSUE THREAD

**Issue by lrupp** [openSUSE/mirrorbrain#28]
Copied from https://github.com/poeml/mirrorbrain/issues/163

I use the .mirrorlist URL a lot in Kali but I also often need to know what mirrors are considered for a given IP.

Right now the only way to approximate this is to add a `?country=XX` query parameter but this obviously does not allow to take into account the AS realm and has the drawback described in https://github.com/poeml/mirrorbrain/issues/80 too.

So please allow usage of `.mirrorlist?ip=X.X.X.X` (or same with IPv6 too) to retrieve the exact mirror list that a given IP would get.

Thank you!

----------

 poeml commented

Too bad, I was about to start hacking and implement this, because it seems to straightforward. However, it's not possible, because mod_mirrorbrain runs after mod_geoip/mod_geoip2/mod_maxminddb, which it relies upon. In other words, at the time mod_mirrorbrain runs it can read the `?ip= parameter` from the URL, but it's too late to look up the GeoIP country and stuff. Thus, it can't be implemented or at least I don't see how.

Under certain circumstances you can still achieve what you asked for. Mod_geoip can respect HTTP headers and read the client up from there (such as is required in a configuration where Apache runs behind a proxy). mod_mirrorbrain looks only at environment variable GEOIP_ADDR, so by that it will also respect the headers that GeoIP read. (With the successor module mod_maxminddb, GEOIP_ADDR is renamed to MMDB_ADDR, which mod_mirrorbrain doesn't know yet.)

Look for 'Proxy-Related Directives' on http://dev.maxmind.com/geoip/legacy/mod_geoip2/ .
Something like
`curl -s -H 'X-Forwarded-For: ...' http://...`
Should then do the job.

**Comment by andrii-suse** [comment]
I think this can be closed: setting up environment variable GEOIP_ADDR or MMDB_ADDR with mod_security will do the job: mod_maxminddb does exactly the same in their tests:
https://github.com/maxmind/mod_maxminddb/blob/master/t/conf/extra.conf.in
```
SecRuleEngine      On
SecRule ARGS:mmdb_addr . "phase:1,t:none,id:32768,log,setenv:'MMDB_ADDR=%{MATCHED_VAR}',msg:'MMDB_ADDR=%{MATCHED_VAR}'
```

**Comment by andrii-suse** [comment]
#48 Includes docker test, which is demonstrating the behavior (the PR includes also fix for switching from GEOIP_ADDR to MMDB_ADDR inside mod_mirrorbrain).
In the test case curl uses interface 127.0.0.2 but is passing GET parameter `mmdb_addr=127.0.0.4`, which becomes used by both mod_maxminddb and mod_mirrorbrain

**Comment by lrupp** [comment]
Perfect :-)

Agreed that with this we can close the issue - but it should also be documented somewhere...
I'm still searching for a good place in  [tree/opensuse/docs](https://github.com/openSUSE/mirrorbrain/)

**Comment by pjessen** [comment]
FWIW, I wrote a wiki entry a while back:
https://progress.opensuse.org/projects/opensuse-admin-wiki/wiki/Which_mirror_to_which_address

**Comment by ott** [comment]
@lrupp Is the documentation referenced @pjessen  sufficient? If so, can you close this issue?

**Comment by andrii-suse** [comment]
Yes X-Forwarded-For should work properly at the moment. E.g. here is list of mirrors for IP of google.com :
```
> curl -H 'X-Forwarded-For: 216.58.215.238' -sL https://download.opensuse.org/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Current.iso.metalink | grep http | head
<metalink version="3.0" xmlns="http://www.metalinker.org/" type="dynamic" origin="https://mirrorcache-us.opensuse.org/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso.metalink" generator="MirrorCache" pubdate="2022-09-22 07:50:06 UTC">
    <url>http://download.opensuse.org</url>
        <url type="https" location="US" preference="100">https://mirrors.ocf.berkeley.edu/opensuse/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="99">https://mirror.clarkson.edu/opensuse/opensuse/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="98">https://mirror.fcix.net/opensuse/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="97">https://mirror.rackspace.com/openSUSE/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="96">https://mirror.sfo12.us.leaseweb.net/opensuse/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="95">https://provo-mirror.opensuse.org/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="94">https://distro.ibiblio.org/opensuse/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
        <url type="https" location="US" preference="93">https://mirrors.rit.edu/opensuse/distribution/leap/15.4/iso/openSUSE-Leap-15.4-DVD-x86_64-Build243.2-Media.iso</url>
```

Adding additional url parameter 'ip=' is good idea as well, because it will not interfere with other services which may look at X-Forwarded-For header. I have it on radar, but feel free to open feature request to MirrorCache as well.
